src/crypto/algorithms/olm.ts

```diff
@@ -209,8 +209,21 @@ class OlmDecryption extends DecryptionAlgorithm {
         //
         // In practice, it's not clear that the js-sdk would behave that way, so this may be only a defence in depth.
 
-        await this.crypto.deviceList.downloadKeys([event.getSender()!], false);
-        const senderKeyUser = this.crypto.deviceList.getUserByIdentityKey(olmlib.OLM_ALGORITHM, deviceKey);
+        let senderKeyUser = this.crypto.deviceList.getUserByIdentityKey(olmlib.OLM_ALGORITHM, deviceKey);
+        if (senderKeyUser === undefined) {
+            // Wait for any pending key query fetches for the user to complete before trying the lookup again.
+            try {
+                await this.crypto.deviceList.downloadKeys([event.getSender()!], false);
+            } catch (e) {
+                throw new DecryptionError("OLM_BAD_SENDER_CHECK_FAILED", "Could not verify sender identity", {
+                    sender: deviceKey,
+                    err: e as Error,
+                });
+            }
+
+            senderKeyUser = this.crypto.deviceList.getUserByIdentityKey(olmlib.OLM_ALGORITHM, deviceKey);
+        }
+
         if (senderKeyUser !== event.getSender() && senderKeyUser != undefined) {
```

Comment:
```suggestion
        if (senderKeyUser !== event.getSender() && senderKeyUser !== undefined) {
```

@richvdh should we also do this ^?

Reply:
probably

Reply:
Oh, hmm. That `senderKeyUser` can be null? Is that why we're not using strict equalities?